src/servers/main.cc

```diff
@@ -212,64 +252,115 @@ SignalHandler(int signum)
   exit_cv_.notify_all();
 }
 
-std::unique_ptr<nvidia::inferenceserver::GRPCServer>
-StartGrpcService(nvidia::inferenceserver::InferenceServer* server)
+nvidia::inferenceserver::Status
+StartMultipleGrpcService(
+    nvidia::inferenceserver::InferenceServer* server,
+    std::map<int32_t, std::vector<std::string>>& port_map)
 {
-  std::unique_ptr<nvidia::inferenceserver::GRPCServer> service;
   nvidia::inferenceserver::Status status =
-      nvidia::inferenceserver::GRPCServer::Create(server, grpc_port_, &service);
+      nvidia::inferenceserver::GRPCServer::Create(
+          server, port_map, &grpc_endpoint_services_);
   if (status.IsOk()) {
-    status = service->Start();
+    for (auto& grpc_eps : grpc_endpoint_services_) {
+      if (grpc_eps != nullptr) {
+        status = grpc_eps->Start();
+      }
+    }
   }
 
   if (!status.IsOk()) {
-    service.reset();
+    for (auto& grpc_eps : grpc_endpoint_services_) {
+      if (grpc_eps != nullptr) {
+        grpc_eps.reset();
+      }
+    }
   }
 
-  return std::move(service);
+  return status;
 }
 
-std::unique_ptr<nvidia::inferenceserver::HTTPServer>
-StartHttpService(nvidia::inferenceserver::InferenceServer* server)
+nvidia::inferenceserver::Status
+StartMultipleHttpService(
+    nvidia::inferenceserver::InferenceServer* server,
+    std::map<int32_t, std::vector<std::string>>& port_map)
 {
-  std::unique_ptr<nvidia::inferenceserver::HTTPServer> service;
   nvidia::inferenceserver::Status status =
       nvidia::inferenceserver::HTTPServer::Create(
-          server, http_port_, http_thread_cnt_, &service);
+          server, port_map, http_thread_cnt_, &http_endpoint_services_);
   if (status.IsOk()) {
-    status = service->Start();
+    for (auto& http_eps : http_endpoint_services_) {
+      if (http_eps != nullptr) {
+        status = http_eps->Start();
+      }
+    }
   }
 
   if (!status.IsOk()) {
-    service.reset();
+    for (auto& http_eps : http_endpoint_services_) {
+      if (http_eps != nullptr) {
+        http_eps.reset();
+      }
+    }
   }
 
-  return std::move(service);
+  return status;
 }
 
 bool
 StartEndpoints(nvidia::inferenceserver::InferenceServer* server)
 {
+  size_t i;
+  std::map<int32_t, std::vector<std::string>> port_map;
```

Comment:
If you declare port_map here isn't it being used for both grpc and http? And then won't it have grpc entries in it when being used for http? Move the declaration down into "allow_grpc" and "allow_http" conditionals.